content/architecture/storage-engine/prolly-tree.md

```diff
@@ -2,38 +2,234 @@
+A Prolly Tree delivers these properties.
+
+Operation | B-Trees | Prolly Trees
+--------- | ------- | ------------
+1 Random Read | log<sub>k</sub>(n) | log<sub>k</sub>(n)
+1 Random Write | log<sub>k</sub>(n) | (1+k/w)*log<sub>k</sub>(n)
+Ordered scan of one item with size z | z/k | z/k
+Calculate diff of size d | n | d
+Structural sharing | ❌ | ✅
+
+**n**: total leaf data in tree, **k**: average block size, **w**: window width
+
+As you can see a Prolly Tree approximates B-tree performance on reads and writes while also offering the ability to compute differences in time proportional to the size of differences rather than the total size of the data. Prolly Trees can also structurally share portions of the tree between versions due to the content-addressed nature of their intermediary storage.
+
+Let's dive a bit deeper and show you how.
+
+# How Prolly Trees work
+
+Prolly Trees are a variant of B-trees so let's first review some key B-tree concepts and then dive into how Prolly Trees work in light of those concepts.
+
+## B-tree Review
+
+A B-tree is data structure that maps keys to values. A B-tree stores key-value pairs in leaf nodes in sorted order. Internal nodes of a B-tree store pointers to children nodes and key delimiters; everything reachable from a pointer to a child node falls within a range of key values corresponding to the key delimiters before and after that pointer within the internal node.
+
+![B-tree Example](../../.gitbook/assets/tim-b-tree-example.png)
+
+A B-tree is optimized for a tradeoff between write and read performance. It's more expensive to maintain than dropping tuples into a heap without any ordering constraints, but when data is stored in a B-tree it's much quicker to seek to a given key and to do an in-order traversal of the keys with their values.
+
+However, finding the differences between two B-trees requires scanning both trees and comparing all values, an operation that scales with the size of the tree. This operation is slow for large trees.
+
+Also, writes to B-trees are not history independent, the order of the writes internally changes the structure of the tree. Thus, storage cannot be easily shared between two versions of the same tree.
+
+## Building a Prolly Tree
+
+The easiest way to understand Prolly trees is to walk through, step-by-step, how we build one. You start with a map of keys to values. Using the above B-tree example, we have keys between 1 and 21. The values don't really matter. Just imagine them dangling off the keys.
+
+1. **Sort**: Sort the map by its key value, so that it is laid out in order. 
+
+![Building a Prolly-Tree, Step 1](../../.gitbook/assets/tim-prolly-tree-step-1.png)
+
+2. **Determine Chunk Boundaries**: Use a seed, the size of the current chunk, and the key value to calculate a rolling hash. Any time the hash value is below a target value, form a chunk boundary and start a new block. Here's the chunking step on our leaf nodes:
```

Comment:
Instead of calling this a rolling hash, you might just say "Use a seed, the size of the current chunk, the key value and a strong hash function to calculate whether this current entry represents a new chunk boundary". Or something like that. It's not technically rolling at this point, just based on the key bytes.